tests/Unit/Http/UrlsTest.php

```diff
@@ -7,24 +7,109 @@
+            ['https://my_user:my_password@some_url.com/path/', 'https://?:?@some_url.com/path/'],
+            ['https://my_user:@some_url.com/path/', 'https://?:@some_url.com/path/'],
+            ['https://my_user:@some_url.com/path/?key=value', 'https://?:@some_url.com/path/'],
+
+            // idempotency
+            ['https://?:@some_url.com/path/?key=value', 'https://?:@some_url.com/path/'],
+            ['?:?@some_url.com/path/?some=value#fragment', '?:?@some_url.com/path/'],
+
+            // false positives that should not be sanitized, but we accept this lack of correctness to reduce complexity
+            ['https://my_user:@some_url.com/before/a:b@/after', 'https://?:@some_url.com/before/?:?@/after'],
+            [
+                'https://my_user:my_passwords@some_url.com/before/a:b@/after',
+                'https://?:?@some_url.com/before/?:?@/after',
+            ],
+        ];
     }
 
-    public function testQueryStringIsRemoved()
+    /**
+     * @dataProvider dataProviderSanitizeDropUserinfo
+     * @param string $url
+     * @param string $expected
+     * @return void
+     */
+    public function testSanitizeDropUserinfo($url, $expected)
     {
-        $this->assertSame('some_url.com/path/', Urls::sanitize('some_url.com/path/?some=value'));
+        /* This test is an exact replica of the same method in tests/Unit/private/UriTest.php and has to be kept in sync
+         * until the other test will be removed as part of the PHP->C migration
+         */
+        $this->assertSame($expected, Urls::sanitize($url, true));
     }
 
-    public function testFragmentIsRemoved()
+    public function dataProviderSanitizeDropUserinfo()
     {
-        $this->assertSame('some_url.com/path/', Urls::sanitize('some_url.com/path/?some=value#fragment'));
+        return [
+            // empty
+            [null, ''],
+            ['', ''],
+
+            // with schema
+            ['https://some_url.com/path/', 'https://some_url.com/path/'],
+
+            // with no schema
+            ['some_url.com/path/', 'some_url.com/path/'],
+
+            // query and fragment
+            ['some_url.com/path/?some=value', 'some_url.com/path/'],
+            ['some_url.com/path/?some=value#fragment', 'some_url.com/path/'],
+
+            // userinfo
+            ['my_user:my_password@some_url.com/path/', 'some_url.com/path/'],
+            ['my_user:@some_url.com/path/', 'some_url.com/path/'],
+            ['my_user:@some_url.com/path/?key=value', 'some_url.com/path/'],
+            ['https://my_user:my_password@some_url.com/path/', 'https://some_url.com/path/'],
+            ['https://my_user:@some_url.com/path/', 'https://some_url.com/path/'],
+            ['https://my_user:@some_url.com/path/?key=value', 'https://some_url.com/path/'],
+
+            // idempotency
+            ['https://?:@some_url.com/path/?key=value', 'https://some_url.com/path/'],
+            ['?:?@some_url.com/path/?some=value#fragment', 'some_url.com/path/'],
```

Comment:
I you truly want idempotency, also add a test without ?query_string part - because the strstr will then skip remove everything starting with the ? in the password.
That's why my regex suggestion wasn't attempting to be idempotent at all, because it hasn't been in the first place.

To be precise, we'd need to strip everything from the _first_ question mark _after_ any potential @.